rkt: add command to remove one or more images from the store.

sgotti opened the pull request:
This should go after #571 and the interesting commits are the last two.

This adds a new command `rkt rmimage` to remove one or more images from the
store providing a key uniquely resolvable.

The logic is to firstly remove transactional data (the rows in the aciinfo and
remote db tables for the given key) and then remove non transactional data (the
diskv blob and imageManifest stores).

The tree store is removed separately and only if an image hash related to the
tree store key isn't referenced by any preparing/prepared/running pod.
This means that an image can be removed (from the db and the blob/imageManifest
stores) but not from the treestore.

As removal of non transactional data can fail for multiple reasons some data
can remain stale, the same applies if the treestore can't be removed because
referenced by a pod.
A future patch will provide a cas gc to clean the stale files and unreferenced
treestores.

Example:

```
$ rkt rmimage sha512-aa244c542b1b631bbd616bc266909b sha512-aaa wronghash
rkt: successfully removed aci for key: "sha512-aa244c542b1b631bbd616bc266909b"
rkt: key "sha512-aaa" not valid: no keys found
rkt: wrong key "wronghash": badly formatted hash string
rkt: 1 image(s) successfully removed
rkt: 2 image(s) cannot be removed
```

sgotti commented:
Updated removing also the treestore.

philips commented on cas/cas.go:
```diff
@@ -264,6 +307,17 @@ func (ds Store) WriteACI(r io.Reader, latest bool) (string, error) {
+	// is different from a lock on an image key as internally
+	// treestore.Write calls the acirenderer functions that use GetACI and
+	// GetImageManifest which are taking the image lock.
+	treestoreLockKey := key + "-treestore"
```
we should have a function to generate this lock name instead of repeating it.

philips replied:
In 8bbbde49c67b927dc2938ea0ea49f73b5d6dd07b we document the semantics of taking the lock. We need to ensure we document those semantics on the function that generates the lock name so a user of the lock knows what she can/cannot do while holding the lock.

jonboulle replied:
+1

sgotti replied:
> we should have a function to generate this lock name instead of repeating it.

@philips Instead of doing this I switched using another lockdir for treestore locks. It seems more logical.

sgotti replied:
> In 8bbbde4 we document the semantics of taking the lock. We need to ensure we document those semantics on the function that generates the lock name so a user of the lock knows what she can/cannot do while holding the lock.

Would you prefer to document the locking semantics in every function or at the start of cas.go?

sgotti replied:
BTW these commits are from #571 (as this PR depends on it). I'm going to update that PR and sync this on top of it. Only the last one is related to this PR.

philips commented on rkt/rmimage.go:
```diff
@@ -0,0 +1,88 @@
+
+var (
+	cmdRmImage = &Command{
+		Name:    "rmimage",
```
With https://github.com/coreos/rocket/pull/649 should we call it `image remove`? /cc @jonboulle @eyakubovich

jonboulle replied:
Probably.... time for subcommands?? should we get @robszumski to help with designing it out?

sgotti replied:
Yeah. Probably it's time for them. Related to #632.

sgotti commented:
Updated with additional checking of tree store keys referenced by any preparing/prepared/running pod. See the updated first PR comment.

krnowak commented on cas/cas.go:
```diff
@@ -57,27 +60,63 @@ var diskvStores = [...]string{
 }
 
+var (
+	ErrKeyNotFound = errors.New("not keys found")
```
Should be "no keys found"

krnowak commented on cas/cas.go:
```diff
@@ -270,7 +335,7 @@ func (ds Store) RenderTreeStore(key string, rebuild bool) error {
 	// This is needed as a previous ACI removal operation could have failed
 	// cleaning the tree store leaving some stale files.
-	err := ds.treestore.Remove(key)
+	err = ds.treestore.Remove(key)
```
You could just do `if err := ds.treestore.Remove(key); err != nil{`

krnowak commented on cas/cas.go:
```diff
@@ -405,6 +486,82 @@ nextKey:
+	}
+	defer treeStoreKeyLock.Close()
+
+	err = ds.treestore.Remove(key)
```
Same here.

krnowak commented on cas/cas.go:
```diff
@@ -405,6 +486,82 @@ nextKey:
+	// remote needs to be removed or a GetRemote will return a blobKey not
+	// referenced by any ACIInfo.
+	err = ds.db.Do(func(tx *sql.Tx) error {
+		found := false
```
This is unnecessary. It is overwritten in next line anyway.

krnowak commented on cas/cas.go:
```diff
@@ -405,6 +486,82 @@ nextKey:
+	// referenced by any ACIInfo.
+	err = ds.db.Do(func(tx *sql.Tx) error {
+		found := false
+		_, found, err := GetACIInfoWithBlobKey(tx, key)
```
You could place it inside an `if`:
```
if _, found, err := Get...(tx,key); err != nil {
	...
} else if !found {
	...
}
```
That limits the scope of found to `if`, `else` only.

krnowak commented on cas/cas.go:
```diff
@@ -405,6 +486,82 @@ nextKey:
+			return fmt.Errorf("cannot find image with key: %s", key)
+		}
+
+		err = RemoveACIInfo(tx, key)
```
You could place it inside `if`.

krnowak commented on cas/cas.go:
```diff
@@ -405,6 +486,82 @@ nextKey:
+		if err != nil {
+			return err
+		}
+		err = RemoveRemote(tx, key)
```
Same here.

krnowak commented on cas/cas.go:
```diff
@@ -405,6 +486,82 @@ nextKey:
+	// leave stale data that will require a cas GC to be implemented.
+	storeErrors := []error{}
+	for _, s := range ds.stores {
+		err := s.Erase(key)
```
Same here.

krnowak commented on cas/cas_test.go:
```diff
@@ -449,5 +449,99 @@ func TestTreeStore(t *testing.T) {
+	key, err := ds.WriteACI(aciFile, false)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
```
The code above is a copypasta from TestGetImageManifest. Consider putting it in common function.

Or maybe postpone it - looks like the ACI preparation in this file really could use some refactoring.

krnowak commented on pkg/lock/keylock.go:
```diff
@@ -0,0 +1,272 @@
+		var lockStat, curStat syscall.Stat_t
+		lfd, err := l.keyLock.Fd()
+		if err != nil {
+			return err
```
At this point keylock is locked. Shouldn't we unlock or close the keylock on error?

sgotti replied:
Right. Thanks. I already fixed it before this review in #763

jonboulle commented:
@sgotti could you please also mention https://github.com/coreos/rkt/pull/774#issuecomment-93950440 ?

sgotti commented:
@krnowak Thanks for your review. I addressed your comments.

@jonboulle As suggested by @iaguis I also added the check on pods in ExitedGarbage state.

Perhaps this is going to wait on #649.

jonboulle commented:
@sgotti per my comment [here](https://github.com/coreos/rkt/pull/587#issuecomment-97592408) please rebase this for merge and we can come back to #649. Thanks!

sgotti commented:
@jonboulle Rebased. Thanks!

jonboulle commented:
thanks, lgtm.
